README
=head1 NAME

ITEMAN Dynamic Publishing - A Perl based dynamic publishing extension for Moveble Type

=head1 DESCRIPTION

  ITEMAN Dynamic Publishing allows you to create dynamic Web pages with normal
  publishing workflow. Since ITEMAN Dynamic Publishing is written in Perl, it
  can be used with any plug-ins.

=head1 SUPPORT

  Visit our website at:
    http://oss.iteman.jp/projects/show/iteman-dynamic-publishing

=head1 COPYRIGHT AND LICENSE

  ITEMAN Dynamic Publishing - A Perl based dynamic publishing extension for Moveble Type
  Copyright (c) 2009 ITEMAN, Inc. All rights reserved.
  
  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


Changes
=head1 NAME

Changes - Release Notes for ITEMAN Dynamic Publishing

=head1 Release Notes for ITEMAN Dynamic Publishing 0.3.0 (beta)

  Release Date: 2009-07-16

  What's New in ITEMAN Dynamic Publishing 0.3.0

   Real Dynamic Publishing:

    In previous versions of ITEMAN Dynamic Publishing, Lazy Static Publishing was only publishing mode. The page what Lazy Static Publishing is applied to, it will be rebuild if required when it is requested by a user.

    In contrast, The page what Real Dynamic Publishing is applied to, it will be rebuild every time when it is requested by a user.

    Since this publishing mode is much slower than Lazy Static Publishing, Use this only if the template really contains dynamic elements.

    For more information, see the Operations Guide.

   License Change:

    The license has been changed from GPLv3 or any later version to GPLv2 or any later version. Why? Real Dynamic Publishing support required to copy and paste a few code of MT::WeblogPublisher included in Movable Type Open Source to our sub-class. And since the license of Movable Type Open Source is GPLv2, it cannot concurrent with GPLv3.

  See the following release notes for details.

  Enhancements
  ============

  - Removed code so that the path of the cache directory and error page for the 404 error is no longer converted to the absolute path to make Movable Type upgrade easier.
  - Changed the behavior of rebuild so that a page which is requested after removing any object is rebuilt.
  - Improved cache management. (Issue #29)
  - Changed the behavior of publish() so that the process continues even though removing the existing page is failed. (lib/ITEMAN/DynamicPublishing.pm)
  - Changed the license to GPLv2 or any later version. (Issue #30)
  - Added support for real dynamic publishing. (Issue #26)
  - Improved error handling.
  - Added code to remove the files and directories which are created by MT for built-in dynamic publishing. (Issue #26)
  - Changed the behavior of clearing caches to not exclude the configuration cache.

  Defect Fixes
  ============

  - Fixed wrong permission for dynamic-publishing.* in the ZIP file.

=head1 Release Notes for ITEMAN Dynamic Publishing 0.2.0 (beta)

  Release Date: 2009-06-09

  What's New in ITEMAN Dynamic Publishing 0.2.0

   * PersistentPerl (SpeedyCGI) and FastCGI support: PersistentPerl (SpeedyCGI) and FastCGI have been supported. You can speed up ITEMAN Dynamic Publishing By using PersistentPerl (SpeedyCGI) or FastCGI. For more information, see the Installation and Configuration at: http://oss.iteman.jp/wiki/iteman-dynamic-publishing/Installation_and_Configuration_Guide
   * Improved performance: The performance of 404 pages has been improved. Also a defect has been fixed that caused performance loss if a file which is not under control of Movable Type is requested.
   * Fallback 404 page: A feature has been added to display the default 404 page when reading the custom 404 page failed.

  See the following release notes for details.

  Enhancements
  ============

  - Removed .objects-save. Now, .objects-save is created by the first access to the "System Plugin Settings" page.
  - Improved messages for the "System Plugin Settings" page.
  - Added error handling for loading the configuration object from the cache.
  - Changed the style of eval blocks again. Because the "IDP: " prefix with $SIG{__DIE__} had not worked. (bin/dynamic-publishing.cgi)
  - Added support for FastCGI. (IssueID 16)
  - Added support for SpeedyCGI (PersistentPerl). (IssueID 23)
  - Improved the performance of 404 pages. (IssueID 24)
  - Added a feature to display the default 404 page when reading the custom 404 page failed. (IssueID 25)

  Defect Fixes
  ============

  - Fixed a defect that caused performance loss if a file which is not under control of Movable Type is requested. (IssueID 22)

=head1 Release Notes for ITEMAN Dynamic Publishing 0.1.0 (beta)

  Release Date: 2009-05-09

  The first release of ITEMAN Dynamic Publishing.


ITEMANDynamicPublishing.pl
# ITEMAN Dynamic Publishing - A Perl-based dynamic publishing system for Moveble Type
# Copyright (c) 2009 ITEMAN, Inc. All rights reserved.
#
# This file is part of ITEMAN Dynamic Publishing.
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

package MT::Plugin::ITEMANDynamicPublishing;

use strict;
use warnings;

use base qw(MT::Plugin);
use ITEMAN::DynamicPublishing::Config;
use ITEMAN::DynamicPublishing::Cache;
use Cwd;

our $VERSION = '0.3.0';

{
    require MT;

    my $description = 'ITEMAN Dynamic Publishing provides a Perl-based dynamic publishing system that works as a replacement for the PHP-based dynamic publishing built into Movable Type. It allows you to publish Web pages dynamically with normal publishing workflow as well as the built-in dynamic publishing.

In contrast to the built-in dynamic publishing, ITEMAN Dynamic Publishing works well with your existing Perl-based plugins.';
    my $author_name = 'ITEMAN, Inc.';
    my $author_link = 'http://iteman.jp/';

    my $plugin = __PACKAGE__->new({
        name => ITEMAN::DynamicPublishing::Config::PLUGIN_NAME,
        id => ITEMAN::DynamicPublishing::Config::PLUGIN_ID,
        key => __PACKAGE__,
        description => "<MT_TRANS phrase=\"$description\">",
        author_name => "<MT_TRANS phrase=\"$author_name\">",,
        author_link => $author_link,
        version => $VERSION,
        system_config_template => 'system.tmpl',
        l10n_class => 'ITEMAN::DynamicPublishing::L10N',
        callbacks => {
            'cms_post_save' => \&update_touch_file,
            'cms_post_delete' => \&update_touch_file,
        },
                                  });
    MT->add_plugin($plugin);

    my $settings = [
        [ 'directory_index', { Default => ITEMAN::DynamicPublishing::Config->default('directory_index'), Scope => 'system'} ],
        [ 'error_page_404', { Default => ITEMAN::DynamicPublishing::Config->default('error_page_404'), Scope => 'system'} ],
        ];
    $plugin->settings(MT::PluginSettings->new($settings));

}

sub update_touch_file {
    require File::Spec;
    require File::Path;

    my $app = MT->instance;
    my @unnecessary_files = (
        File::Spec->catfile($app->blog->site_path, '.htaccess'),
        File::Spec->catfile($app->blog->site_path, 'mtview.php'),
        File::Spec->catfile($app->blog->site_path, 'templates_c'),
        File::Spec->catfile($app->blog->site_path, 'cache'),
        );
    if ($app->blog->archive_path) {
        push @unnecessary_files, File::Spec->catfile($app->blog->archive_path, '.htaccess');
        push @unnecessary_files, File::Spec->catfile($app->blog->archive_path, 'mtview.php');
    }
    foreach (@unnecessary_files) {
        if (-e $_) {
            if (-d $_) {
                File::Path::rmtree($_);
            } else {
                unlink $_;
            }
        }
    }

    utime undef, undef, (ITEMAN::DynamicPublishing::Config::REBUILD_TOUCH_FILE)
        or die "Failed to change the access and modification times on " . ITEMAN::DynamicPublishing::Config::REBUILD_TOUCH_FILE;
}

sub load_config {
    my $plugin = shift;
    my ($param, $scope) = @_;

    $plugin->SUPER::load_config(@_);

    unless (ITEMAN::DynamicPublishing::Cache->new->load('ITEMAN::DynamicPublishing::Config')) {
        eval { $plugin->_save_idp_config($param) };
    }

    unless (-f ITEMAN::DynamicPublishing::Config::REBUILD_TOUCH_FILE) {
        eval {
            open REBUILD_TOUCH_FILE, '> ' . ITEMAN::DynamicPublishing::Config::REBUILD_TOUCH_FILE or die;
        };
        unless ($@) {
            close REBUILD_TOUCH_FILE;
        }
    }

    $param->{cache_directory} = ITEMAN::DynamicPublishing::Config::CACHE_DIRECTORY;

    eval {
        $plugin->_check_cache_directory($param);
    };
    if ($@) {
        $param->{cache_directory_error} = 1;
    }

    eval {
        $plugin->_check_error_page_404($param);
    };
    if ($@) {
        $param->{error_page_404_error} = 1;
    }

    $param->{is_configured} = ITEMAN::DynamicPublishing::Cache->new->load('ITEMAN::DynamicPublishing::Config') && !$param->{cache_directory_error} && !$param->{error_page_404_error} ? 1 : 0;
}

sub save_config {
    my $plugin = shift;
    my ($param, $scope) = @_;

    unless ($scope eq 'system') {
        return;
    }

    if ($param->{clear_caches} eq 'true') {
        ITEMAN::DynamicPublishing::Cache->new->clear;
        return;
    }

    if ($param->{directory_index} eq '') {
        return $plugin->error($plugin->translate('Directory Index is required'));
    }

    if ($param->{error_page_404} eq '') {
        return $plugin->error($plugin->translate('The error page for the status code 404 is required'));
    }

    $plugin->SUPER::save_config(@_);

    $plugin->_save_idp_config($param);
}

sub reset_config {
    my $plugin = shift;

    $plugin->SUPER::reset_config(@_);

    ITEMAN::DynamicPublishing::Cache->new->clear;
}

sub _save_idp_config {
    my $plugin = shift;
    my $param = shift;

    my $config = ITEMAN::DynamicPublishing::Config->new;
    $config->directory_index($param->{directory_index});
    $config->error_page_404($param->{error_page_404});
    $config->db_dsn($MT::Object::DRIVER->fallback->dsn);
    $config->db_user($MT::Object::DRIVER->fallback->username);
    $config->db_password($MT::Object::DRIVER->fallback->password);

    ITEMAN::DynamicPublishing::Cache->new->save({
        cache_id => 'ITEMAN::DynamicPublishing::Config',
        data => $config,
                                                });
}

sub _check_cache_directory {
    my $plugin = shift;
    my $param = shift;

    my $user_name = getpwuid $>;
    $user_name = $> unless $user_name;
    unless (-d $param->{cache_directory}) {
        $param->{cache_directory_error_message} = $plugin->translate('The directory is not found or not readable for [_1]', $user_name);
        die;
    }

    unless (-x $param->{cache_directory}) {
        $param->{cache_directory_error_message} = $plugin->translate('The directory is not readable for [_1]', $user_name);
        die;
    }

    unless (-w $param->{cache_directory}) {
        $param->{cache_directory_error_message} = $plugin->translate('The directory is not writable for [_1]', $user_name);
        die;
    }
}

sub _check_error_page_404 {
    my $plugin = shift;
    my $param = shift;

    my ($feature) = (caller(1))[3] =~ /(error_page_\d+)$/;

    my $user_name = getpwuid $>;
    $user_name = $> unless $user_name;
    unless (-f $param->{error_page_404}) {
        $param->{error_page_404_error_message} = $plugin->translate('The file is not found or not readable for [_1]', $user_name);
        die $param->{error_page_404_error_message};
    }

    unless (-r $param->{error_page_404}) {
        $param->{error_page_404_error_message} = $plugin->translate('The file is not readable for [_1]', $user_name);
        die $param->{error_page_404_error_message};
    }
}

1;

# Local Variables:
# mode: perl
# coding: iso-8859-1
# tab-width: 4
# c-basic-offset: 4
# c-hanging-comment-ender-p: nil
# indent-tabs-mode: nil
# End:
